File: Data Structures & Algorithms/max-area-of-island/submission-0.cpp
class Solution {
public:
    int dfs(vector<vector<int>>& grid, int i, int j) {
        
        int n = grid.size();
        int m = grid[0].size();

        if(i < 0 || j < 0 || i >= n || j >= m)
            return 0;

        if(grid[i][j] == 0)
            return 0;

        grid[i][j] = 0;

        return 1
            + dfs(grid, i+1, j)
            + dfs(grid, i-1, j)
            + dfs(grid, i, j+1)
            + dfs(grid, i, j-1);
    }
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int maxArea = 0;
        int n = grid.size();
        int m = grid[0].size();
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {

                if(grid[i][j] == 1) {
                    int area = dfs(grid, i, j);
                    maxArea = max(maxArea, area);
                }
            }
        }

        return maxArea;
    }
};

File: Data Structures & Algorithms/swim-in-rising-water/submission-0.java
class Solution {

    boolean dfs(int limit, int[][] grid, boolean[][] visited, int i, int j) {
        int n = grid.length;
        int m = grid[0].length;

        if (i < 0 || i >= n || j < 0 || j >= m ||
            visited[i][j] || grid[i][j] > limit) {
            return false;
        }

        if (i == n - 1 && j == m - 1) {
            return true;
        }

        visited[i][j] = true;

        return dfs(limit, grid, visited, i - 1, j) ||
               dfs(limit, grid, visited, i + 1, j) ||
               dfs(limit, grid, visited, i, j - 1) ||
               dfs(limit, grid, visited, i, j + 1);
    }

    public int swimInWater(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;

        int l = Integer.MAX_VALUE;
        int h = Integer.MIN_VALUE;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                l = Math.min(l, grid[i][j]);
                h = Math.max(h, grid[i][j]);
            }
        }

        while (l < h) {
            int mid = l + (h - l) / 2;

            boolean[][] visited = new boolean[n][m];

            if (dfs(mid, grid, visited, 0, 0)) {
                h = mid;
            } else {
                l = mid + 1;
            }
        }

        return l;
    }
}